File: symfony/src/Service/Marketplace/CardPriceHistoryService.php
<?php

namespace App\Service\Marketplace;

use App\Entity\Marketplace\CardPriceHistory;
use App\Repository\Marketplace\CardPriceHistoryRepository;

class CardPriceHistoryService
{
    public function __construct(
        private CardPriceHistoryRepository $repository,
    ) {}

    public function create(array $data): CardPriceHistory
    {
        throw new \LogicException('Not implemented');
    }

    public function update(CardPriceHistory $entity, array $data): CardPriceHistory
    {
        throw new \LogicException('Not implemented');
    }

    public function priceChangePercent(int $id, $previousAvg): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('CardPriceHistory not found: ' . $id);
        $result = $entity->priceChangePercent($previousAvg);
        $this->repository->save($entity, flush: true);
        return $result;
    }

    public function isPriceSpike(int $id, $thresholdPercent): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('CardPriceHistory not found: ' . $id);
        $result = $entity->isPriceSpike($thresholdPercent);
        $this->repository->save($entity, flush: true);
        return $result;
    }
}

File: rails/app/services/marketplace/card_price_history_service.rb
module Marketplace
  class CardPriceHistoryService
    def initialize(repository = CardPriceHistory)
      @repository = repository
    end

    def create(attributes)
      raise NotImplementedError, "#{self.class}#create not implemented"
    end

    def update(entity, attributes)
      raise NotImplementedError, "#{self.class}#update not implemented"
    end

    def price_change_percent(id)
      instance = CardPriceHistory.find(id)
      result = instance.price_change_percent(previous_avg)
      instance.save!
      result
    end

    def is_price_spike(id)
      instance = CardPriceHistory.find(id)
      result = instance.is_price_spike(threshold_percent)
      instance.save!
      result
    end
  end
end
